addons/hr/models/hr_employee_departure.py

```diff
@@ -0,0 +1,204 @@
+    def _compute_apply_immediately(self):
+        today = fields.Date.today()
+        for departure in self:
+            if departure.state in ['done', 'cancel']:
+                departure.apply_immediately = False
+            elif departure.action_at == "departure_date":
+                departure.apply_immediately = departure.departure_date <= today
+            elif departure.action_at == "other":
+                departure.apply_immediately = not departure.action_other_date or departure.action_other_date <= today
+            else:
+                departure.apply_immediately = False
+
+    @api.depends('action_at', 'departure_date', 'action_other_date')
+    def _compute_apply_date(self):
+        today = fields.Date.today()
+        for departure in self:
+            if departure.action_at == 'departure_date':
+                departure.apply_date = max(departure.departure_date, today)
+            else:
+                departure.apply_date = departure.action_other_date or today
+
+    def _check_departure_validity(self):
+        for departure in self:
+            active_version = departure.employee_id.current_version_id
+            if active_version.contract_date_start and active_version.contract_date_start > departure.departure_date:
+                raise ValidationError(self.env._("Departure date can't be earlier than the start date of current contract."))
+
+    @api.model_create_multi
+    def create(self, vals_list):
+        res = super().create(vals_list)
+        for departure in res:
+            departure.employee_id._get_version(departure.departure_date).write({
+                'departure_id': departure.id,
+            })
+        return res
+
+    def action_schedule(self):
+        self.state = 'scheduled'
+
+    def action_cancel(self):
+        self.state = 'cancelled'
+        self.employee_id.version_id.departure_id = False
+
+    def _cron_apply_departure(self):
+        departures = self.search([
+            ('state', '=', 'scheduled'),
+            ('apply_date', '<=', fields.Date.today()),
+        ])
+        for departure in departures:
+            departure.action_register()
+
+    def action_register(self):
+        def _get_user_archive_notification_action(message, message_type, next_action):
+            return {
+                'type': 'ir.actions.client',
+                'tag': 'display_notification',
+                'params': {
+                    'title': self.env._("User Archive Notification"),
+                    'type': message_type,
+                    'message': message,
+                    'next': next_action,
+                },
+            }
+
+        departures_per_user = {}
+        for departure in self:
+            if departure.employee_id.user_id in departures_per_user:
+                departures_per_user[departure.employee_id.user_id] += departure
+            else:
+                departures_per_user[departure.employee_id.user_id] = departure
```

Comment:
```suggestion
        departures_per_user = departure.grouped(lambda departure: departure.employee_id.user_id)
```
I think this is clearer,  what do you think ? :D